solidity/contracts/ERC1155MixedFungibleMintable.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.6.0;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+
+/**
+    @dev Mintable form of ERC1155 with mixed fungible/non-fungible item support.
+    Based on reference implementation:
+    https://github.com/enjin/erc-1155/blob/master/contracts/ERC1155MixedFungibleMintable.sol
+*/
+contract ERC1155MixedFungibleMintable is Context, ERC1155 {
+    // Use a split bit implementation:
+    //   - Bit 255: type flag (0 = fungible, 1 = non-fungible)
+    //   - Bits 254-128: pool id
+    //   - Bits 127-0: token index (non-fungible only)
+    uint256 constant TYPE_MASK = uint256(uint128(~0)) << 128;
+    uint256 constant NF_INDEX_MASK = uint128(~0);
+    uint256 constant TYPE_NF_BIT = 1 << 255;
+
+    uint256 nonce;
+    mapping (uint256 => address) public creators;
+    mapping (uint256 => uint256) public maxIndex;
+
+    function isFungible(uint256 id) internal pure returns(bool) {
+        return id & TYPE_NF_BIT == 0;
+    }
+    function isNonFungible(uint256 id) internal pure returns(bool) {
+        return id & TYPE_NF_BIT == TYPE_NF_BIT;
+    }
+
+    // Only the creator of a token type is allowed to mint it.
+    modifier creatorOnly(uint256 type_id) {
+        require(creators[type_id] == _msgSender());
+        _;
+    }
+
+    constructor(string memory uri) ERC1155(uri) public {
+    }
+
+    function create(string calldata uri, bool is_fungible)
+        external
+        virtual
+        returns(uint256 type_id)
+    {
+        type_id = (++nonce << 128);
+        if (!is_fungible)
+          type_id = type_id | TYPE_NF_BIT;
+
+        creators[type_id] = _msgSender();
+
+        emit TransferSingle(_msgSender(), address(0x0), address(0x0), type_id, 0);
+
+        if (bytes(uri).length > 0)
+            emit URI(uri, type_id);
```

Comment:
Realized that since the code is now coupled closely to this specific contract, there's no need to abuse the base `URI` event to encode parameters from FireFly. I'm planning to add a custom `TokenCreate` event in this contract instead, and pad in an arbitrary `data` parameter instead of a URI.